onedal/datatypes/data_conversion_sua_iface.cpp

```diff
@@ -0,0 +1,157 @@
+
+    dal::table res{};
+
+    if (is_readonly) {
+        res = dal::homogen_table(queue,
+                                 ptr,
+                                 r_count,
+                                 c_count,
+                                 deleter,
+                                 std::vector<sycl::event>{},
+                                 layout);
+    }
+    else {
+        auto* const mut_ptr = const_cast<Type*>(ptr);
+        res = dal::homogen_table(queue,
+                                 mut_ptr,
+                                 r_count,
+                                 c_count,
+                                 deleter,
+                                 std::vector<sycl::event>{},
+                                 layout);
+    }
+    obj.inc_ref();
+    return res;
+}
+
+dal::table convert_from_sua_iface(py::object obj) {
+    auto sua_iface_dict = get_sua_interface(obj);
+    const auto type = get_sua_dtype(sua_iface_dict);
+
+    dal::table res{};
+
+#define MAKE_HOMOGEN_TABLE(CType) res = convert_to_homogen_impl<CType>(obj);
+
+    SET_DAL_TYPE_FROM_DAL_TYPE(type,
+                               MAKE_HOMOGEN_TABLE, //
+                               report_problem_for_sua_iface(": unknown data type"));
+
+#undef MAKE_HOMOGEN_TABLE
+
+    return res;
+}
+
+py::dict construct_sua_iface(const dal::table& input) {
+    const auto kind = input.get_kind();
+    if (kind != dal::homogen_table::kind())
+        report_problem_to_sua_iface(": only homogen tables are supported");
+
+    const auto& homogen_input = reinterpret_cast<const dal::homogen_table&>(input);
+    const dal::data_type dtype = homogen_input.get_metadata().get_data_type(0);
+    const dal::data_layout data_layout = homogen_input.get_data_layout();
+
+    npy_intp row_count = dal::detail::integral_cast<npy_intp>(homogen_input.get_row_count());
+    npy_intp column_count = dal::detail::integral_cast<npy_intp>(homogen_input.get_column_count());
+
+    // need "version", "data", "shape", "typestr", "syclobj"
+    py::tuple shape = py::make_tuple(row_count, column_count);
+    py::list data_entry(2);
+
+    auto bytes_array = dal::detail::get_original_data(homogen_input);
+    auto has_queue = bytes_array.get_queue().has_value();
+    // OneDAL returns tables without sycl context for CPU sycl queue inputs, that
+    // breaks the compute-follows-data execution.
+    // Currently not throwing runtime exception and __sycl_usm_array_interface__["syclobj"] None asigned
+    // if no Sycl queue to allow workaround on python side.
+    // if (!has_queue) {
+    //     report_problem_to_sua_iface(": table has no queue");
+    // }
+
+    const bool is_mutable = bytes_array.has_mutable_data();
+
+    static_assert(sizeof(std::size_t) == sizeof(void*));
+    data_entry[0] = is_mutable ? reinterpret_cast<std::size_t>(bytes_array.get_mutable_data())
+                               : reinterpret_cast<std::size_t>(bytes_array.get_data());
+    data_entry[1] = is_mutable;
```

Comment:
I don't know whats going on here unfortunately, could you add some comments to clarify?  what is element 1 of data_entry, is it a SUA standard telling if it can be modified or not?

Reply:
Comments with references are added.